libr/bin/format/mach0/coresymbolication.c

```diff
@@ -0,0 +1,306 @@
+				break;
+			}
+			sect->size = r_read_ble (cursor, false, bits);
+			cursor += word_size;
+			if (cursor >= end) {
+				break;
+			}
+			ut64 sect_name_off = r_read_ble (cursor, false, bits);
+			cursor += word_size;
+			if (bits == 32) {
+				cursor += word_size;
+			}
+			ut8 *sect_name = sect_start + sect_name_off;
+			if (sect_name >= end) {
+				break;
+			}
+			r_str_ncpy (sect->name, (const char *)sect_name, 256);
+		}
+	}
+	ut64 start_of_symbols = start_of_sections +
+			(ut64)hdr->n_sections * (bits == 32 ? R_CS_EL_SIZE_SECT_32 : R_CS_EL_SIZE_SECT_64);
+	if (hdr->n_symbols) {
+		result->symbols = R_NEWS0 (RCoreSymCacheElementSymbol, hdr->n_symbols);
+		if (!result->symbols) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_symbols;
+		for (i = 0; i < hdr->n_symbols && cursor + R_CS_EL_SIZE_SYM <= end; i++) {
+			RCoreSymCacheElementSymbol *sym = &result->symbols[i];
+			sym->paddr = r_read_le32 (cursor);
+			sym->size = r_read_le32 (cursor + 0x4);
+			sym->unk1 = r_read_le32 (cursor + 0x8);
+			ut64 name_off = r_read_le32 (cursor + 0xc);
+			ut64 mangled_name_off = r_read_le32 (cursor + 0x10);
+			sym->unk2 = (st32)r_read_le32 (cursor + 0x14);
+
+			if (cursor + name_off < b || cursor + name_off >= end) {
+				break;
+			}
+			r_str_ncpy (sym->name, (const char *)cursor + name_off, 512);
+
+			if (cursor + mangled_name_off < b || cursor + mangled_name_off >= end) {
+				break;
+			}
+			r_str_ncpy (sym->mangled_name, (const char *)cursor + mangled_name_off, 512);
+
+			cursor += R_CS_EL_SIZE_SYM;
+		}
+	}
+	ut64 start_of_lined_symbols = start_of_symbols + (ut64)hdr->n_symbols * R_CS_EL_SIZE_SYM;
+	if (hdr->n_lined_symbols) {
+		result->lined_symbols = R_NEWS0 (RCoreSymCacheElementLinedSymbol, hdr->n_lined_symbols);
+		if (!result->lined_symbols) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_lined_symbols;
+		for (i = 0; i < hdr->n_lined_symbols && cursor + R_CS_EL_SIZE_LSYM <= end; i++) {
+			RCoreSymCacheElementLinedSymbol *lsym = &result->lined_symbols[i];
+			lsym->sym.paddr = r_read_le32 (cursor);
+			lsym->sym.size = r_read_le32 (cursor + 0x4);
+			lsym->sym.unk1 = r_read_le32 (cursor + 0x8);
+			ut64 name_off = r_read_le32 (cursor + 0xc);
+			ut64 mangled_name_off = r_read_le32 (cursor + 0x10);
+			lsym->sym.unk2 = (st32)r_read_le32 (cursor + 0x14);
+			ut64 file_name_off = r_read_le32 (cursor + 0x18);
+			lsym->flc.line = r_read_le32 (cursor + 0x1c);
+			lsym->flc.col = r_read_le32 (cursor + 0x20);
+
+			if (cursor + name_off < b || cursor + name_off >= end) {
+				break;
+			}
+			r_str_ncpy (lsym->sym.name, (const char *)cursor + name_off, 512);
+
+			if (cursor + mangled_name_off < b || cursor + mangled_name_off >= end) {
+				break;
+			}
+			r_str_ncpy (lsym->sym.mangled_name, (const char *)cursor + mangled_name_off, 512);
+
+			if (cursor + file_name_off < b || cursor + file_name_off >= end) {
+				break;
+			}
+			r_str_ncpy (lsym->flc.file, (const char *)cursor + file_name_off, 512);
```

Comment:
same, dont harccode buffer size. and imho it shuoldnt be chopping or leaking  unused memory

Reply:
what if its not null terminated? the length cant be 512 b ecause it can perform an oobread .

I would suggest to use r_buf apis everywhere for safety and clean api. but you shuold know how many bytes left are there from `cursor` to the end of the buffer that is pointing to